### PR TITLE
worked design proposal for supervised forecasting

### PR DESCRIPTION
**Pull request by fkiraly**
worked design proposal for supervised forecasting

@partmor, @mloning

### CODE REVIEW

File: steps/04_supervised_forecasting_api/step.md [hunk 1]
```diff
@@ -1,15 +1,165 @@
...
 First models to implement:
+
 * reduction to supervised tabular
 * mixed effects models
 * reduction 2-step: use only constant variables for mean then another estimator for forecast
 
 Other resources
+
 * Master thesis (to be uploaded)
 * previous design notes (will see if I can find any)
```
Comment (mloning) [inline steps/04_supervised_forecasting_api/step.md:165]: We can remove this, you mention all of these points in more detail above
```suggestion
```

**Review by mloning — COMMENTED** [review]
Looks good to me, happy to merge this with one minor comment

**Comment by mloning** [comment]
One question: what's the reason for using fit(X, cutoffs=None, ...) over fit(X, Y=None, ...)?

**Comment by fkiraly** [comment]
>One question: what's the reason for using fit(X, cutoffs=None, ...) over fit(X, Y=None, ...)?

What would `Y` stand for, here?
If you suggest a simple rename, then I would oppose this on the grounds of user expectation: `Y` is the thing to be predicted, usually - the cut-offs are not the thing to be predicted though.

File: steps/04_supervised_forecasting_api/step.md (hunk 1 above)
Comment (fkiraly) [reply steps/04_supervised_forecasting_api/step.md:165]: sure - just thought to keep it for completeness

**Comment by fkiraly** [comment]
>I will play around with some Python code when I find some time to try out different things.

You also need to leave some of the playing to @partmor :-)

File: steps/04_supervised_forecasting_api/step.md (hunk 1 above)
Comment (mloning) [reply steps/04_supervised_forecasting_api/step.md:165]: But they will be in the git history, no?

**Comment by mloning** [comment]
I was thinking of `Y` being the actual data for the "target" time window, so coming from something like this:
```python
X, Y = data[:, :, :cutoff], data[:, :, cutoff:]
```
This splits the data outside of the estimator rather than inside as in your proposal. I'm just wondering what the pros and cons are for each?

**Comment by mloning** [comment]
One other suggestion, I've come to prefer the term panel forecasting over supervised forecasting, panel forecasting seems to be widely used already, aligns with our panel as the name for the data format, and is a bit shorter too.

**Comment by fkiraly** [comment]
>This splits the data outside of the estimator rather than inside as in your proposal. I'm just wondering what the pros and cons are for each?

My reason for that was:
* in general there does not need to be a designated cut-off - for example, if all the input series are samples from a (panel) ARMA process (or considered as such).
* the "y" predicted in prediction mode, specified by `fh`, does in general not need to agree with the "y" in the training set.

The cut-off time or index may also be different for the observed series - not sure whether we want to cover that case though.

**Comment by fkiraly** [comment]
>One other suggestion, I've come to prefer the term panel forecasting over supervised forecasting, panel forecasting seems to be widely used already, aligns with our panel as the name for the data format, and is a bit shorter too.

I think the two are different tasks!

In panel forecasting, to the best of my knowledge, you try to forecast the future of the entire panel. While you assume that there are i.i.d. instances, there is no "training" or "test" set, and the pasts/futures of all the instances align. This can be seen as a special case of multivariate or vector forecasting, with an additional i.i.d. assumption across variables.

The problem covered in this PR is closer to supervised regression in the tabular setting - "past" taking the role of "X" and "future" of "y". 
Individual "pasts" and "futures" across samples may differ, e.g., in the case where relative to a trigger date (e.g., therapy), you have already observed some patients' future (in the training set), while for others you haven't and are looking for a forecast (test/prediction set).

Now it also may be worth to cover panel forecasting, but in my opinion not before we have solved multivariate forecasting properly - since it is more closely related to that.

Quick question: given your view of literature, do you think the above is an accurate reflection? Or am I confusing terms?

**Comment by partmor** [comment]
Hi all, I've been away due to unforeseen issues. I'm getting back on the wagon with your updates and discussions. Will be back to you!

**Pull request merged**